src/pydantic_pint/quantity.py

```diff
@@ -143,7 +150,9 @@ def validate(
             if isinstance(v, Number) and self.restriction == "dimensions" and not self.strict:
                 raise ValueError("must specify units")
             if isinstance(v, Quantity) and self.restriction == "units":
-                v = v.to(self.units, *self.ureg_contexts)
+                _v, v = v, v.to(self.units, *self.ureg_contexts)
+                if v.magnitude != _v.magnitude and self.exact:
```

Comment:
To benefit from python's short-circuit evaluation (bool evaluates faster than magntude comparison), this could be changed to:

```suggestion
                _v, v = v, v.to(self.units, *self.ureg_contexts)
                if self.exact and v.magnitude != _v.magnitude:
```

Additionally, to further speed up the evaluation and since destination `units` is known, it should be possible to check for equality of units (saving the conversion into self.units):

```suggestion
                if self.exact and v.units != self.units.units:
```

Reply:
Agreed on both. Fixed in [6e742f4](https://github.com/tylerh111/pydantic-pint/pull/13/commits/6e742f4293602aa41ce670961a810fc057ef2e85). One note, the conversion `v.to(self.units, ...)` still needs to happen for the normal case, but that can be done after the check. And because we have the units stored, we can just check against that, instead of holding on to the old quantity.